File: content-solutions/graph-advanced/topo-build-time.js
function solution(times, prereqs) {
  const n = times.length;
  const graph = Array.from({ length: n }, () => []);
  const indeg = new Array(n).fill(0);
  for (const [a, b] of prereqs) {
    graph[a].push(b);
    indeg[b]++;
  }
  const finish = [...times];
  const queue = [];
  for (let v = 0; v < n; v++) if (indeg[v] === 0) queue.push(v);
  for (let head = 0; head < queue.length; head++) {
    const v = queue[head];
    for (const w of graph[v]) {
      finish[w] = Math.max(finish[w], finish[v] + times[w]);
      if (--indeg[w] === 0) queue.push(w);
    }
  }
  return Math.max(...finish);
}

File: content-solutions/graph-advanced/topo-build-time.java
import java.util.*;

class Solution {
    public int solution(int[] times, int[][] prereqs) {
        int n = times.length;
        List<List<Integer>> graph = new ArrayList<>();
        for (int i = 0; i < n; i++) graph.add(new ArrayList<>());
        int[] indeg = new int[n];
        for (int[] p : prereqs) {
            graph.get(p[0]).add(p[1]);
            indeg[p[1]]++;
        }
        int[] finish = times.clone();
        ArrayDeque<Integer> queue = new ArrayDeque<>();
        for (int v = 0; v < n; v++) if (indeg[v] == 0) queue.add(v);
        while (!queue.isEmpty()) {
            int v = queue.poll();
            for (int w : graph.get(v)) {
                finish[w] = Math.max(finish[w], finish[v] + times[w]);
                if (--indeg[w] == 0) queue.add(w);
            }
        }
        int answer = 0;
        for (int f : finish) answer = Math.max(answer, f);
        return answer;
    }
}
